src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+      break;
+    }
+
+    DebugAssert(!node->right_input(), "Scan only implemented for nodes with one input");
+    node = node->left_input();
+  }
+
+  // Remove projections/aggregates found below the last predicate which we don't need to remove/patch.
+  info.projection_nodes.resize(num_projections_to_remove);
+  info.aggregate_nodes.resize(num_aggregates_to_patch);
+
+  return info;
+}
+
+/**
+ * Check whether an LQP node uses a correlated parameter.
+ *
+ * If the node uses a subquery, its nodes are also all checked.
+ */
+template <class ParameterPredicate>
+bool uses_correlated_parameters(const std::shared_ptr<AbstractLQPNode>& node,
+                                ParameterPredicate&& is_correlated_parameter) {
+  bool is_correlated = false;
+  for (const auto& expression : node->node_expressions) {
+    visit_expression(expression, [&](const auto& sub_expression) {
+      // We already know that the node is correlated, so we can skip the rest of the expression
+      if (is_correlated) {
+        return ExpressionVisitation::DoNotVisitArguments;
+      }
+
+      if (sub_expression->type == ExpressionType::LQPSubquery) {
+        // Need to check whether the subquery uses correlated parameters
+        const auto& lqp_select_expression = std::static_pointer_cast<LQPSubqueryExpression>(sub_expression);
+        visit_lqp(lqp_select_expression->lqp, [&](const auto& sub_node) {
+          if (is_correlated) {
+            return LQPVisitation::DoNotVisitInputs;
+          }
+
+          is_correlated |= uses_correlated_parameters(sub_node, is_correlated_parameter);
+          return is_correlated ? LQPVisitation::DoNotVisitInputs : LQPVisitation::VisitInputs;
+        });
+      } else if (sub_expression->type == ExpressionType::CorrelatedParameter) {
+        const auto& parameter_expression = std::static_pointer_cast<CorrelatedParameterExpression>(sub_expression);
+        is_correlated |= is_correlated_parameter(parameter_expression->parameter_id);
+      }
+
+      return is_correlated ? ExpressionVisitation::DoNotVisitArguments : ExpressionVisitation::VisitArguments;
+    });
+
+    if (is_correlated) {
+      break;
+    }
+  }
+
+  return is_correlated;
+}
+
+/**
+ * Searches for usages of correlated parameters that we cannot optimize.
+ *
+ * This includes two things:
+ *   - Usages of correlated parameters outside of predicate nodes (for example joins)
+ *   - Usages of correlated parameters in predicate nodes nested below nodes they cannot be pulled up past.
+ */
+template <class ParameterPredicate>
+bool contains_unoptimizable_correlated_parameter_usages(
+    const std::shared_ptr<AbstractLQPNode>& lqp, ParameterPredicate&& is_correlated_parameter,
+    const std::set<std::shared_ptr<PredicateNode>> optimizable_predicate_nodes) {
+  bool optimizable = true;
+  visit_lqp(lqp, [&](const auto& node) {
+    if (!optimizable) {
+      return LQPVisitation::DoNotVisitInputs;
```

Comment:
When would this be called?

Reply:
As in the case above, this is to stop the visitation as fast as possible once we have found something that prevents optimization.